### ISSUE TITLE
Exception in SQL generation

### ISSUE THREAD

**Issue by LorenzBuehmann** [ontop/ontop#381]
Hi again,

getting an exception with the following setup:

Ontop 4.0.3

Database:

```sql
CREATE TABLE IF NOT EXISTS "t11" (s varchar(255) NOT NULL, o varchar(255) NOT NULL)
CREATE TABLE IF NOT EXISTS "t12" (s varchar(255) NOT NULL, o varchar(255) NOT NULL, l varchar(10))
CREATE TABLE IF NOT EXISTS "t2" (s varchar(255) NOT NULL, o varchar(255) NOT NULL)
CREATE TABLE IF NOT EXISTS "t3" (s varchar(255) NOT NULL, o LONG NOT NULL)
```

Mappings:

```
[MappingDeclaration] @collection [[
mappingId     "m12"
source        SELECT "s", "o" FROM "t12" WHERE  "l" = 'en' 
target        <{s}> <http://ex.org/p1>  "{o}"@en  .



mappingId     "m2"
source        SELECT "s", "o" FROM "t2"
target        <{s}> <http://ex.org/p2> <{o}> .



mappingId     "m3"
source        SELECT "s", "o" FROM "t3"
target        <{s}> <http://ex.org/p3> "{o}"^^<http://www.w3.org/2001/XMLSchema#integer> .



mappingId     "m11"
source        SELECT "s", "o" FROM "t11"
target        <{s}> <http://ex.org/p1> <{o}> .

]]
```

SPARQL:
```sparql
SELECT  *
WHERE
  { ?s  <http://ex.org/p1>  ?o ;
        <http://ex.org/p2>  ?o1 ;
        ?p                  ?o
  }
LIMIT   10
```

Exception:
```
20/12/29 09:16:30 WARN QuestQueryProcessor: Unexpected exception: Multiple entries with same key: o2m1=V7."o2m1" and o2m1=V3."o2m1"
java.lang.IllegalArgumentException: Multiple entries with same key: o2m1=V7."o2m1" and o2m1=V3."o2m1"
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:146)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:109)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:394)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:582)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:230)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SQLNaryJoinExpressionImpl.acceptVisitor(SQLNaryJoinExpressionImpl.java:33)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.getSQLSerializationForChild(DefaultSelectFromWhereSerializer.java:260)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:62)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SelectFromWhereWithModifiersImpl.acceptVisitor(SelectFromWhereWithModifiersImpl.java:99)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.lambda$visit$11(DefaultSelectFromWhereSerializer.java:238)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:239)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SQLUnionExpressionImpl.acceptVisitor(SQLUnionExpressionImpl.java:35)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.getSQLSerializationForChild(DefaultSelectFromWhereSerializer.java:260)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:62)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SelectFromWhereWithModifiersImpl.acceptVisitor(SelectFromWhereWithModifiersImpl.java:99)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.getSQLSerializationForChild(DefaultSelectFromWhereSerializer.java:254)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:62)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SelectFromWhereWithModifiersImpl.acceptVisitor(SelectFromWhereWithModifiersImpl.java:99)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.lambda$visit$11(DefaultSelectFromWhereSerializer.java:238)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:239)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SQLUnionExpressionImpl.acceptVisitor(SQLUnionExpressionImpl.java:35)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.getSQLSerializationForChild(DefaultSelectFromWhereSerializer.java:260)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:62)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SelectFromWhereWithModifiersImpl.acceptVisitor(SelectFromWhereWithModifiersImpl.java:99)
	at net.sansa_stack.query.spark.ontop.SparkSelectFromWhereSerializer.serialize(SparkSelectFromWhereSerializer.java:39)
	at it.unibz.inf.ontop.generation.DefaultSQLIQTree2NativeNodeGenerator.generate(DefaultSQLIQTree2NativeNodeGenerator.java:52)
	at it.unibz.inf.ontop.answering.reformulation.generation.impl.SQLGeneratorImpl.generateNativeNode(SQLGeneratorImpl.java:140)
	at it.unibz.inf.ontop.answering.reformulation.generation.impl.SQLGeneratorImpl.generateSourceQuery(SQLGeneratorImpl.java:84)
	at it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.generateExecutableQuery(QuestQueryProcessor.java:148)
	at it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.reformulateIntoNativeQuery(QuestQueryProcessor.java:121)
```

Observations when I tried to reproduce it from a larger example:
* It does not occur if the type of column `t3.o` resp. the literal in mapping `m3` have the same type as `t12.o` resp. the literal in `m12`, i.e. a string literal in the current example
* It also does not occur if only `t11`/`m11` or `t12`/`m12` exist, but not both
* It does not occur if the second triple pattern of the query (`?s <http://ex.org/p2>  ?o1`) will be removed

So it looks like it has something to do with 
- having two mappings that generate the same property with different types (URI and literal)
- and another mapping with a different literal datatype
- and the query has to have two triple patterns with bound predicates and one triple pattern with unbound predicate and shared object variable

Can you confirm this issue? 

Thanks in advance

**Comment by bcogrel** [comment]
Hi Lorenz,

Thanks a lot for this bug report. We managed to reproduce it and to fix it (https://github.com/ontop/ontop/commit/9149041f0c6efee3c631b947b12d850a324d49e9). The fix will be released with 4.1.0.

In a nutshell, Ontop is replacing co-occurrences of variables by explicit equalities just before generating the SQL query. The bug was happening during this normalization.

Best,
Benjamin

**Comment by LorenzBuehmann** [comment]
Ah cool. Thanks for fast response and fix - as usual by your team. Appreciate it.

Cheers,
Lorenz
